Adds basic documentation

samgdotson opened the pull request:
This PR closes #7 by adding basic documentation to the repository. It includes
- contrib.md
- detailed readme file
- docs directory for sphinx documentation

samgdotson committed:
adds extras requirements for docs

samgdotson committed:
adds initial commit for documentation

samgdotson committed:
adds contrib file

samgdotson committed:
Merge branch 'main' of github.com:arfc/osier into docs

pep8speaks commented:
Hello @samgdotson! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





There are currently no PEP 8 issues detected in this Pull Request. Cheers! :beers: 

##### Comment last updated at 2022-09-13 20:09:34 UTC

samgdotson committed:
pep8 to docs pr

samgdotson committed:
adds pronunciation to readme

samgdotson committed:
pep8 in setup.py

samgdotson committed:
improves instructions in contrib

samgdotson committed:
adds sphinx extensions

samgdotson committed:
fixes sphinx extension list

samgdotson committed:
pep8 setup.py

yardasol commented on README.md:
```diff
@@ -1,2 +1,44 @@
+
+## Installation
+
+`osier` is not currently on PyPI or Conda-forge (we hope this will change
```
I'd avoid trying to make promises in your documentation unless you are 100% sure it will happen. For you though since one of the requirements of naming this software was that it can be put on PyPI, I think this is okay since you seriously intend to put it up there. Conda-forge will be a little trickier, so perhaps remove that clause for now?

samgdotson replied:
(I am 100% sure it will happen :))

yardasol commented on README.md:
```diff
@@ -1,2 +1,44 @@
+# for a basic installation
+pip install .
+# to also build the documentation
+pip install .[doc]
```
I tried running this command but the docs did not build.

samgdotson replied:
This installs optional dependencies, info to build the docs is a little farther down. Though I see how this could be confusing.

yardasol commented on docs/source/conf.py:
```diff
@@ -0,0 +1,64 @@
+extensions = [
+    "myst_parser",
+    "numpydoc",
+    "sphinxcontrib.napoleon",
```
```suggestion
    "sphinx.ext.napoleon",
```
When running `make html` after installing `osier` with Python 3.10, there is an error:

```python
Extension error:
Could not import extension sphinxcontrib.napoleon (exception: cannot import name 'Callable' from 'collections' (/home/ooblack/miniconda3/envs/osier-env/lib/python3.10/collections/__init__.py))
```

It turns out that [Python 3.10 moved the `Callable` object from the `collections` module to the `typing` module](https://docs.python.org/3/whatsnew/3.10.html#pep-612-parameter-specification-variables). Looking at the [PyPI page for `sphinxcontrib-napoleon`](https://pypi.org/project/sphinxcontrib-napoleon/), we can see that this package is now included with `sphinx` after version 1.3 as `sphinx.ext.napoleon`. Making this change, the build runs successfully.

yardasol replied:
Just a heads up, there are two warning messages I get when running `make html` after making this change:

```
WARNING: html_static_path entry '_static' does not exist
```
and
```
/home/ooblack/projects/osier/README.md:44: WARNING: 'myst' reference target not found: docs/source/contrib.md
```

[Here](https://gist.github.com/yardasol/fcb04b0187902fb780412c765da807ba) is the full output

samgdotson replied:
Thanks for the `napoleon` update! Regarding the two warnings you receive, they're both ignorable for now. 
* The _static folder doesn't exist because I don't have any static files and git won't let you push an empty directory. 
* The `myst` reference warning is a bit strange and may require a separate issue. The explanation is that I prefer to have a single contrib file rather than a contrib file in the docs that references another contributing file. I'm not sure what the best way to handle this is, the only difference is that the README has a hyperlink to the contrib file when viewed on github, while the docs has a hyperlink immediately below the reference.

yardasol commented on setup.py:
```diff
@@ -50,16 +50,29 @@
+    'doc': [
+        'sphinx',
+        'myst-parser',
+        'sphinxcontrib-napoleon',
```
```suggestion
```

yardasol commented on docs/source/contrib.md:
```diff
@@ -0,0 +1,52 @@
+
+The [GitHub "Issues" tab](https://github.com/arfc/osier/issues) contains open issues. Some issues
+are labeled "Difficulty:1-Beginner" and are good first issues to tackle for new contributers to 
+`osier`. You can even sort by "label" to find precisely those issues!
```
```suggestion
We use [GitHub "Issues" tab](https://github.com/arfc/osier/issues) to track bugs and feature requests. Issues
labeled "Difficulty:1-Beginner" are good first issues for new contributors to tackle.
You can sort by "label" to find these issues.
```
Some sentence structure suggestions. I don't want to take your voice out of the writing, so feel free to ignore this particular restructuring.

samgdotson replied:
This is fine, thanks!

yardasol commented on docs/source/contrib.md:
```diff
@@ -0,0 +1,52 @@
+`osier`. You can even sort by "label" to find precisely those issues!
+
+## Bug reports
+Bug reports from users are an important for identifying unforeseen issues. If you find a bug please
+open a new issue with the following:
```
```suggestion
Bug reports from users are important for identifying unintentional erroneous behavior in the code. If you find a bug, please
open a new issue with the following:
```

yardasol commented on docs/source/contrib.md:
```diff
@@ -0,0 +1,52 @@
+## Bug reports
+Bug reports from users are an important for identifying unforeseen issues. If you find a bug please
+open a new issue with the following:
+1. An explanation of the problem with enough details for others to reproduce the problem. Some 
+common information needed is
```
```suggestion
1. An explanation of the problem with enough details for others to reproduce the problem. You should include the following information at a minimum
```

samgdotson replied:
It's not a minimum requirement because some issues might not need all that information! For example, the code might run without error, but the results don't make sense.

yardasol commented on README.md:
```diff
@@ -1,2 +1,44 @@
+## Contributing
+
+Contributions to `osier` are welcome. For details on how to make bug reports, issues to
+work on, and other information, check the [contributing page](docs/source/contrib.md)
```
```suggestion
Contributions to `osier` are welcome. For details on how to make bug reports, pull requests, and other information, check the [contributing page](docs/source/contrib.md)
```

I think this makes a bit more sense, as issues to work on are in the issues tab on the github repo, not at the contributing page in the documentation.

yardasol reviewed (changes requested):
Hi Sam, this is a great start.

Since this is a new project, consider using a [`setup.cfg` file](https://docs.python.org/3/distutils/configfile.html). A good example of what this looks like in practice can be found in [this PR](https://github.com/openmc-dev/openmc/pull/2144/files#diff-fa602a8a75dc9dcc92261bac5f533c2a85e34fcceaff63b3a3a81d9acde2fc52)

My other comments are below.

samgdotson committed:
Apply suggestions from @yardasol

Co-authored-by: Olek <45364492+yardasol@users.noreply.github.com>

samgdotson committed:
resolves other comments from @yardasol

samgdotson committed:
Merge branch 'docs' of github.com:samgdotson/osier into docs

samgdotson commented:
@yardasol Thanks for the review! I'm a little hesitant to add a setup.cfg file, only because it replicates information already found in `setup.py`.  [nrelpy](https://pypi.org/project/nrelpy/) is published on PyPI and doesn't have a `setup.cfg`, so it isn't strictly necessary from a packaging perspective. Why do you recommend it?

Otherwise, I think I've addressed all of your comments and it should be ready to merge.

samgdotson commented:
Bump @yardasol. 

@nsryan2 Do you want to take a look?

nsryan2 commented on docs/source/contrib.md:
````diff
@@ -0,0 +1,52 @@
+```
+The `-e` flag creates an editable installation so you don't have to reinstall everytime a
+feature is changed.
+
````
```suggestion

### For MacOS Users
If you encounter an error
`zsh: no matches found: .[doc]`
running the final pip installation command for the docs, other users have remedied the error by running:

`pip install -e .'[doc]'`
```

nsryan2 replied:
The suggestion was getting a little wonky when I tried to use 
```bash
code
````
 hence the single ``.

nsryan2 commented on docs/source/contrib.md:
```diff
@@ -0,0 +1,52 @@
+Bug reports from users are important for identifying unintentional behavior. If you find a bug, please
+open a new issue with the following:
+1. An explanation of the problem with enough details for others to reproduce the problem. Some 
+common information needed is
```
```suggestion
common information needed is:
```

nsryan2 commented on docs/source/contrib.md:
```diff
@@ -0,0 +1,52 @@
+    * Python version
+    * Any commands executed (perhaps a python snippet)
+    * An error message from the terminal
+2. An explanation of the expected behavior. For example
```
```suggestion
2. An explanation of the expected behavior. For example:
```

nsryan2 commented on docs/source/contrib.md:
```diff
@@ -0,0 +1,52 @@
+
+## Making a pull request
+A good pull request requires the following, along with a new feature (where applicable)
+1. All functions should have docstrings using the [Numpydoc style](https://numpydoc.readthedocs.io/en/latest/format.html)
```
```suggestion
1. All functions should have docstrings using the [Numpydoc style](https://numpydoc.readthedocs.io/en/latest/format.html).
```

nsryan2 commented on README.md:
```diff
@@ -1,2 +1,43 @@
+
+## Contributing
+
+Contributions to `osier` are welcome. For details on how to make bug reports, pull requests, and other information, check the [contributing page](docs/source/contrib.md)
```
```suggestion
Contributions to `osier` are welcome. For details on how to make bug reports, pull requests, and other information, check the [contributing page](docs/source/contrib.md).
```

nsryan2 reviewed (approved):
We were able to get everything up and running! Everything looks to be in a good place, just a few periods missing. I tried suggesting the fix you found for Mac with the zsh error, but the suggestion got a little wonky with embedded code blocks (just a heads up).

samgdotson committed:
Apply suggestions from @nsryan2

Co-authored-by: Nathan Ryan <nathansean0@gmail.com>

samgdotson commented:
@nsryan2 Thanks for the suggestions! If you think it works, then you can merge at will.
